libass/ass_parse.c

```diff
@@ -1010,6 +1010,16 @@ void process_karaoke_effects(ASS_Renderer *render_priv)
                         x_end = d6_to_int(e1->pos.x + e1->advance.x);
                         dt = (tm_current - tm_start);
                         dt /= (tm_end - tm_start);
+                        frz = fmod(s1->frz, 360);
+                        if (frz > 90 && frz < 270) {
+                            // Fill from right to left
+                            dt = 1 - dt;
+                            for (cur2 = s1; cur2 <= e1; ++cur2) {
+                                uint32_t tmp = cur2->c[0];
+                                cur2->c[0] = cur2->c[1];
```

Comment:
If we can redefine colors here then for uniformity we should handle `tm_current < tm_start` and `tm_current >= tm_end` through the same trick and eliminate usages of `info->effect_type` in `render_text()`.

Reply:
Ironically, I’m actually not very happy with my solution of fudging the colours here.

On the other hand, I don’t immediately see any harm in what you’re suggesting.

Furthermore, it might help us move closer to a solution to another long-standing problem: the edge between the glyph body and the ASS border sometimes looks bad because the body and border are separate bitmaps and one is drawn on top of the other. I have [an old set of patches](https://github.com/astiob/libass/commits/alpha) that fixes this, but it doesn’t work with karaoke. The reason is: it needs to know the body’s alpha value when it’s rasterizing the border. But karaoke can change the body’s alpha, and it does this _way_ after the border has been rasterized. So if we were to make karaoke change the colour here, that would help! We’d only need to figure out what to do with \kf.

Reply:
In my tiled rendering experiment I've transformed all sorts of clipping into generic multiply by bitmap. I think something in that line should be done for karaoke too.